crates/bevy_ecs/macros/src/lib.rs

```diff
@@ -423,6 +424,411 @@ pub fn derive_system_param(input: TokenStream) -> TokenStream {
     })
 }
 
+static READONLY_ATTRIBUTE_NAME: &str = "readonly";
+static FILTER_ATTRIBUTE_NAME: &str = "filter";
+
+/// Implement `WorldQuery` to use a struct as a parameter in a query
+#[proc_macro_derive(Fetch, attributes(readonly, filter))]
+pub fn derive_fetch(input: TokenStream) -> TokenStream {
```

Comment:
I'd prefer to see this split out into a seperate file within the same folder. This file is getting quite long, and there's very little shared code / logic between the various macros.